alphagov/paas-cf
# [#155854402] Adapt internal certificates to cf deployment bosh variables

[#155854402 Restructure internal certs to match cf-deployment](https://www.pivotaltracker.com/story/show/155854402)

## What

We want to start using certificates as they do in the [cf-deployment](https://github.com/cloudfoundry/cf-deployment):
 - Use the same naming convention
 - With [certs generated using bosh variables](https://bosh.io/docs/variable-types.html)

Of course, this change must:
 - Be without downtime.
 - Still allow rotate the certificates without downtime.

In this PR we will:
 - rename all the certificates as in cf-deployment
 - start using bosh-interpolate variable generation (state kept in a separated `cf-vars-certs.yml` file)
 - Implement logic to migrate from the existing `cf-certs.tgz` to the new `cf-vars-certs.yml`
 - Adapt the certificate rotation scripts to rotate the certs in `cf-vars-certs.yml`
 - Allow rotate also the dedicated CA certs without downtime.
 - Use the rotation of CAs and certs to stop using the bosh CA everywhere, and start with dedicated CA certs as in cf-deployment

## Implementation details

### About certificate rotation

We implement certificate rotation as follows: 

 - All CA certs have two variables `<cert_name>` and `<cert_name>_old`. In the manifest we use both CA certs concatenated.
 - To rotate, we follow these 3 steps:
    1. Rename the old CA certs from `<cert_name>` to `<cert_name>_old`.
    2. Generate new CA certs.
    2. Deploy. Both new and old CA are used everywhere.
    3. Generate new certificates for all the other certs.
    4. Deploy. Old certs and new certs would work because both old and new CAs are deployed.
    5. Delete all the `_old` certs.
    6. Deploy.

### About the bosh-CA to dedicated-CA migration

 - When migrating from `cf-certs.tgz` to `cf-vars-store.yml`, we will use the `bosh-CA` as the value of every single CA.
 - Then we will do a normal certificate rotation: Rotate CAs, deploy both, generate new certs, deploy, and delete the old ones.

## How to review

The first time you run this branch it shall be almost a noop except:
 - The `.instance_groups[diego-cell].jobs[rep].properties.diego.rep.server_cert` and `.instance_groups[diego-cell].jobs[rep].properties.diego.rep.server_key` change. 
   This is because now we use the same certificate for the `rep` client and server. 
 - `.instance_groups[router].jobs[gorouter].properties.router.ca_certs` is added. This is to allow gorouter to validate the certificate of UAA.

First deployment:

 1. Deploy from `master`
 2. Copy `aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-manifest.yml cf-manifest-v1.yml ` that we ill compare later
 3. pause `cf-deploy` 
 4. Deploy from this branch up to `generate-cf-config.
 5. Compare the manifests:
     -  `aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-manifest.yml cf-manifest-v2.yml `
     - `spruce diff cf-manifest-v1.yml cf-manifest-v2.yml  | less -R`

 6. Unpause the pipeline. All test should pass including availability test.

Once that is deployed, we must do a full certificate rotation:

 1. Run the job `credentials > rotate-cf-certs-cas`. 
     - This rename existing CA certs to `_old` in the `vars-store`. New ones will be generate in next deploy.
 2. Run all the `create-cloudfoundry:all`  pipeline. 
     - There should be no downtime (maybe a pair of errors in API availability test).
     - New CA certs would be generated by `bosh interpolate`
     - All jobs using CAs would be updated to include both the old and new CA certs.
 3. Run the job `credentials > rotate-cf-certs-leafs`. 
     - It will delete the old certs in the `vars-store`
 4. Run all the `create-cloudfoundry:all`  pipeline. 
     - There should be no downtime (maybe a pair of errors in API availability test).
     - New CA certs would be generated by `bosh interpolate`
     - All jobs using certs would be updated with the env certs
 5. Run the job `credentials > delete-old-cf-certs`. 
     - It will delete the old certs in the `vars-store`.
 4. Run all the `create-cloudfoundry:all`  pipeline. 
     - There should be no downtime. 
     - All the old certs (CAs or not) would be deleted.

To test the certificate expire job in concourse:

 1. Update the number of days to check in the job `create-cloudfoundry.yml` / `check-var-store-certs`, from 30 to 340. 
 2. Push the pipeline and rerun the job.
 3. Check that a mail has been sent to: https://groups.google.com/a/digital.cabinet-office.gov.uk/forum/#!forum/govpaas-alerting-dev

## Follow up.

 - Deploy this in prod
 - Rotate certs in prod, following the procedure above.
 - Remove the migrate job `cf-certs` job in a different PR (FIXME: do other PR)

## Who can review

Anyone but @camelpunch or @keymon or @dcarley
